Hunter-420/DevFlexPort
# feat/forgot-password

# Development of Forgot Password Feature

## Overview:
This pull request introduces the implementation of a "Forgot Password" feature with two new APIs: `forgot-password` and `reset-password`. The `forgot-password` API allows users to request a one-time password (OTP) via email for resetting their password. The `reset-password` API allows users to reset their password using a valid OTP.

## Changes Made:
- Added a new route and controller (`forgotPassword`) for the `forgot-password` API.
  - Validates the user's email address.
  - Generates a 6-digit OTP and sends it to the user's email.
  - Sets an expiration time (10 minutes) for the OTP and stores it in the user's document.
  - Uses `nodemailer` to send the OTP via email.

- Added a new route and controller (`resetPassword`) for the `reset-password` API.
  - Validates the OTP provided by the user.
  - Checks if the OTP is valid and has not expired.
  - Hashes and updates the user's password with the new password provided by the user.
  - Clears the OTP from the user's document after a successful password reset.

## How to Test:
1. **Forgot Password (forgot-password API)**:
   - Send a POST request to `/api/forgot-password` with the user's email address.
   - Check your email inbox for the OTP.
   - Use the OTP to reset the password within 10 minutes.

2. **Reset Password (reset-password API)**:
   - Send a POST request to `/api/reset-password` with the user's email, OTP, and new password.
   - Ensure that the OTP is valid and has not expired.
   - Verify that the password is successfully updated.

## Additional Notes:
- Implemented error handling to handle cases such as invalid OTPs or expired OTPs.
- Used `bcrypt` for password hashing to securely update user passwords.
- Included logging and debugging statements for troubleshooting.

## Changes Required in Production:
Please ensure to update the `.env` file in the production environment with the following variables:
```plaintext
EMAIL_ADDRESS='your-EMAIL_ADDRESS-here'
EMAIL_PASSWORD='your-EMAIL_PASSWORD-here'
```
These variables are required for sending emails (using nodemailer) in the forgotPassword API.

- #### Note:
  For Gmail users, ensure that "Less Secure App Access" is enabled in your Google account settings to allow sending emails programmatically. Follow [these instructions](https://support.google.com/accounts/answer/6010255) to enable it.

## Checklist:
- [x] Implemented forgot-password and reset-password APIs.
- [x] Added error handling and input validation.
- [x] Integrated email functionality for sending OTPs.
- [x] Tested locally and ensured functionality.

## Reviewer/Maintainer Tasks:
- [ ] Update the .env file in the production environment with the required email credentials.
- [ ] Test the feature in the production environment to ensure proper functionality.
